### PR TITLE
sql: descriptor bookkeeping for column families

### PR DESCRIPTION
**Pull request by danhhz**
This commit adds the proto representations of column families. They are filled
in new tables and existing tables are backfilled when read from kv or received
from gossip. The column families are _unused_ in sql, but all the necessary
validation is present. This means that the bookkeeping to keep the family fields
correct through all updates and codepaths should be here.

NB: The backfilled descriptors may be written to kv, but this is okay because
old code will ignore the new fields and they do not yet affect correctness.
Everything is still backward and forward compatible.

Followup commits will handle migrating data in a forward-only way and then
continue the column family implementation. This is when the FormatVersion will
be bumped.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7089)

### CODE REVIEW

File: sql/lease.go [hunk 1]
```diff
@@ -152,6 +152,7 @@ func (s LeaseStore) Acquire(
 		return nil, util.Errorf("ID %d is not a table", tableID)
 	}
 
+	tableDesc.MaybeUpgradeFormatVersion()
```
Comment (tschottdorf) [inline sql/lease.go]: does `tableDesc` need to be validated here as well?

**Comment by paperstreet** [comment]
cc @vivekmenezes if you want to poke at the schema change aspect of this. The existing tests helped me catch the most obvious edge cases. It turned out not so complex since we don't support changing the family of a column as a schema change (yet). As Validate implies, the invariant I went with was for each column (including all those in mutations) is assigned to a column family. The upcoming implementation is responsible for doing the correct thing with the columns in a family.

File: sql/lease.go (hunk 1 above)
Comment (paperstreet) [inline sql/lease.go]: line 158 does it (TableDescriptor is embedded in LeaseState)

**Comment by tschottdorf** [comment]
Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089)*

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[sql/sqlbase/structured.go, line 266 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJgkqe1ldE6jxhluDlj:-KJgkqe2Z3IJcf2tK2IY:-1281460853) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.go#L266)):*
> ```Go
> // FormatVersion (if it's not already there) and returns true if any changes
> // were made.
> func (desc *TableDescriptor) MaybeUpgradeFormatVersion() bool {
> ```

Maybe instead of using/checking versions, simply populate the family info if it's not there?

---

*[sql/sqlbase/structured.proto, line 67 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJglG6tnDF_KtgrR94C:-KJglG6tnDF_KtgrR94D:1442646754) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.proto#L67)):*
> ```protobuf
> }
> 
> message FamilyDescriptor {
> ```

deserves a comment (its not even clear these are families of columns)

---

*[sql/sqlbase/structured.proto, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJgmsg5sW7unK1dYFDB:-KJgmsg5sW7unK1dYFDC:749621418) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.proto#L76)):*
> ```protobuf
>   // names here proves to be prohibitive, we could clear this field before
>   // saving and reconstruct it after loading.
>   repeated string column_names = 3;
> ```

how necessary is this field? seems like the column ids should be enough

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089)*

**Comment by vivekmenezes** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[sql/schema_changer.go, line 620 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJkHFQP2JuTnrynALot:-KJkHFQP2JuTnrynALou:-854726716) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/schema_changer.go#L620)):*
> ```Go
> 					case *sqlbase.Descriptor_Table:
> 						table := union.Table
> 						table.MaybeUpgradeFormatVersion()
> ```

I don't see a need for this. This notification is merely a hint that something in the table descriptor might have changed warranting  an execution of a schema change. Unless you need it for Validate?

---

*[sql/sqlbase/structured.proto, line 67 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJglG6tnDF_KtgrR94C:-KJkF1ieKtMNLck1dSZV:-1253992884) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.proto#L67)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> deserves a comment (its not even clear these are families of columns)
</details>
I'd call it ColumnFamily. Not sure if you need the descriptor suffix ?

---

*[sql/sqlbase/structured.proto, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJgmsg5sW7unK1dYFDB:-KJkFMlWCvDGiWom-ehZ:5172523) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.proto#L76)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> how necessary is this field? seems like the column ids should be enough
</details>
+1

---

*[sql/sqlbase/structured.proto, line 90 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJkG81u0kWTB4mP1sYK:-KJkG81u0kWTB4mP1sYL:272432735) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.proto#L90)):*
> ```protobuf
>   // so this field supplies it.
>   optional uint32 default_column_id = 5 [(gogoproto.nullable) = false,
>       (gogoproto.customname) = "DefaultColumnID", (gogoproto.casttype) = "ColumnID"];
> ```

Is this really necessary? Can you use the first column id in column_ids? or store a boolean indicating that the default is the first column_id?  or store the index into the column_ids?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089)*

**Comment by paperstreet** [comment]
Thanks for the reviews! RFAL

<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> cc @vivekmenezes if you want to poke at the schema change aspect of this. The existing tests helped me catch the most obvious edge cases. It turned out not so complex since we don't support changing the family of a column as a schema change (yet). As Validate implies, the invariant I went with was for each column (including all those in mutations) is assigned to a column family. The upcoming implementation is responsible for doing the correct thing with the columns in a family.
</details>


---

Review status: 5 of 10 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[sql/schema_changer.go, line 620 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJkHFQP2JuTnrynALot:-KJkOmkaMX_YGn4nn1vo:-1615441239) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/schema_changer.go#L620)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I don't see a need for this. This notification is merely a hint that something in the table descriptor might have changed warranting  an execution of a schema change. Unless you need it for Validate?
</details>
Yeah, it's needed for Validate. Fortunately, after I can fix up the reference version tests and do the TODO about saving back the migrated version, it'll almost always be a no-op.

---

*[sql/sqlbase/structured.go, line 266 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJgkqe1ldE6jxhluDlj:-KJkQQkH1Yurcs1jUEf2:1668376564) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.go#L266)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Maybe instead of using/checking versions, simply populate the family info if it's not there?
</details>
The versions are needed to prevent old versions from being started on data that has started to use column families. I think in the long run, this code is easier to reason about if you just have to think about transitions from version to version so that's what I was planning on once the TODO below is done.

The interim is an interesting question that I went back and forth on. I had it checking for empty families first and it worked well, but since this works now (it just overwrites the families every time, which is fine since they're deterministically derivable from the rest of the descriptor) I didn't see a good reason not to use it. Willing to revisit

---

*[sql/sqlbase/structured.proto, line 67 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJglG6tnDF_KtgrR94C:-KJkUw5NZq9uYOVTVyRE:-264744306) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.proto#L67)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I'd call it ColumnFamily. Not sure if you need the descriptor suffix ?
</details>
Everything else is suffixed by Descriptor, so I'll keep that for consistency. Added the Column prefix and a comment.

---

*[sql/sqlbase/structured.proto, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJgmsg5sW7unK1dYFDB:-KJkUYkAXPbRdSI_KaHe:-91973448) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.proto#L76)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> +1
</details>
I don't like it either, but we'd have to change how parsing and AllocateIDs work and I've got enough yaks to shave in this project already so I'm trying to contain scope. Plus, IndexDescriptor does it the same way, so if we fix it, we can fix both and delete the fields.

---

*[sql/sqlbase/structured.proto, line 90 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJkG81u0kWTB4mP1sYK:-KJkV3yI9fZ_EyLqOnTw:-238141304) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.proto#L90)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Is this really necessary? Can you use the first column id in column_ids? or store a boolean indicating that the default is the first column_id?  or store the index into the column_ids?
</details>
Unfortunately, I can't see how to do it without another field or something much hackier because of a couple wrinkles. First, there's a difference between a family that once had a default set (it had one field, then another was later added) and one that hasn't (it was created with two fields). Second, the column that was the default could later be deleted.

Storing the ColumnID is the most explict and the easiest to keep correct

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-:-KJkYu74GgwUu2k4SmKB:-124023199)*

**Comment by vivekmenezes** [comment]
:lgtm:

<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Thanks for the reviews! RFAL
</details>


---

Review status: 5 of 10 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[sql/sqlbase/structured.go, line 766 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJlOvbQg5NKpWgTmuvH:-KJlOvbRJTwo4vsDnpqG:-1490356994) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.go#L766)):*
> ```Go
> 			desc.Families[i].ColumnNames = append(
> 				desc.Families[i].ColumnNames[:found], desc.Families[i].ColumnNames[found+1:]...)
> 			break
> ```

How about moving this code inside the for loop and returning instead, and not using found

---

*[sql/sqlbase/structured.go, line 772 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJlPJlZJMz4VWuvUu0S:-KJlPJl_DVp9mC8LspSa:243306635) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.go#L772)):*
> ```Go
> 
> // RenameColumn updates all references to a column name in indexes and families.
> func (desc *TableDescriptor) RenameColumn(colID ColumnID, newColName string) {
> ```

Can you create an issue to test this code. I suppose you want to do that once you have the sql implementation in place.

---

*[sql/sqlbase/structured.proto, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJgmsg5sW7unK1dYFDB:-KJlKWEhTTFssIqCi-zb:345410218) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.proto#L76)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> I don't like it either, but we'd have to change how parsing and AllocateIDs work and I've got enough yaks to shave in this project already so I'm trying to contain scope. Plus, IndexDescriptor does it the same way, so if we fix it, we can fix both and delete the fields.
</details>
Yes we should keep it. I forgot that a column being assigned to a group is done through SQL. We try to contain everything provided by SQL in one place. Sounds fine. No need to add a TODO to cleanup.

---

*[sql/sqlbase/structured.proto, line 90 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJkG81u0kWTB4mP1sYK:-KJlLIsuffiiugkHDI23:5173568) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.proto#L90)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Unfortunately, I can't see how to do it without another field or something much hackier because of a couple wrinkles. First, there's a difference between a family that once had a default set (it had one field, then another was later added) and one that hasn't (it was created with two fields). Second, the column that was the default could later be deleted.
> 
> Storing the ColumnID is the most explict and the easiest to keep correct
</details>
LG

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-:-KJlPjOLQY3BlTZasek7:1417088721)*

**Comment by RaduBerinde** [comment]
:lgtm:

<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> :lgtm:
</details>


---

Review status: 5 of 10 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-:-KJlSNU0BkC3UY9ac2pN:1417088721)*

**Comment by paperstreet** [comment]
Review status: 5 of 10 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[sql/sqlbase/structured.go, line 766 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJlOvbQg5NKpWgTmuvH:-KJlRupu8RBQSFl8Ef8y:18049019) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.go#L766)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> How about moving this code inside the for loop and returning instead, and not using found
</details>
Wow, yeah. No idea why it did it that way. This is much better.

---

*[sql/sqlbase/structured.go, line 772 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089#-KJlPJlZJMz4VWuvUu0S:-KJlRzV-TX3RBph6KJgz:-1032731851) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c258ced02e26870b0eb01cf6ba5a2fdf60b5692/sql/sqlbase/structured.go#L772)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Can you create an issue to test this code. I suppose you want to do that once you have the sql implementation in place.
</details>
This is all checked by fairly well by Validate, which is called quite a bit in the course of running Logic Tests. I added an issue to be more direct about it: https://github.com/cockroachdb/cockroach/issues/7112

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7089)*

**Pull request merged**